File: backend/app/utils/mini_program_uri.py
"""Parser for the shared mini-program URI profile."""

from __future__ import annotations

import re
from dataclasses import dataclass
from urllib.parse import unquote

MINI_PROGRAM_URI_SCHEME = "miniprogram"
_PREFIX = f"{MINI_PROGRAM_URI_SCHEME}://"
_ACTIONS = {"navigate-to"}
_PATH_SEGMENT = re.compile(r"^(?:[A-Za-z0-9._~!$&'()*+,;=:@-]|%[0-9A-Fa-f]{2})+$")
_QUERY = re.compile(r"^(?:[A-Za-z0-9._~!$&'()*+,;=:@/?-]|%[0-9A-Fa-f]{2})*$")


@dataclass(frozen=True)
class MiniProgramUri:
    action: str
    route: str


def parse_mini_program_uri(value: str) -> MiniProgramUri | None:
    """Parse a URI after validating its raw, non-normalized components."""

    raw = value.strip()
    if raw[: len(_PREFIX)].lower() != _PREFIX or "#" in raw:
        return None

    hierarchy_and_query = raw[len(_PREFIX) :]
    query_index = hierarchy_and_query.find("?")
    if query_index >= 0:
        hierarchy = hierarchy_and_query[:query_index]
        raw_query: str | None = hierarchy_and_query[query_index + 1 :]
    else:
        hierarchy = hierarchy_and_query
        raw_query = None

    path_index = hierarchy.find("/")
    if path_index <= 0:
        return None

    action = hierarchy[:path_index].lower()
    if action not in _ACTIONS:
        return None

    raw_path = hierarchy[path_index:]
    if raw_path == "/" or not raw_path.startswith("/"):
        return None
    if raw_query is not None and _QUERY.fullmatch(raw_query) is None:
        return None

    segments = raw_path[1:].split("/")
    if any(_PATH_SEGMENT.fullmatch(segment) is None for segment in segments):
        return None

    for segment in segments:
        try:
            decoded = unquote(segment, encoding="utf-8", errors="strict")
        except (UnicodeDecodeError, ValueError):
            return None
        if (
            decoded in {".", ".."}
            or "/" in decoded
            or "\\" in decoded
            or "\0" in decoded
        ):
            return None

    query_suffix = "" if raw_query is None else f"?{raw_query}"
    return MiniProgramUri(action=action, route=f"{raw_path}{query_suffix}")

File: frontend/src/utils/miniProgramUri.ts
export const MINI_PROGRAM_URI_SCHEME = 'miniprogram';

export type MiniProgramUriAction = 'navigate-to';

export type MiniProgramUri = {
    action: MiniProgramUriAction;
    /** Raw mini-program route, including the leading slash and optional query. */
    route: string;
};

const PREFIX = `${MINI_PROGRAM_URI_SCHEME}://`;
const ACTIONS = new Set<MiniProgramUriAction>(['navigate-to']);
const PATH_SEGMENT = /^(?:[A-Za-z0-9._~!$&'()*+,;=:@-]|%[0-9A-Fa-f]{2})+$/;
const QUERY = /^(?:[A-Za-z0-9._~!$&'()*+,;=:@/?-]|%[0-9A-Fa-f]{2})*$/;

/**
 * Parse the platform mini-program environment URI profile.
 *
 * The profile follows RFC 3986's hierarchical URI shape:
 * `miniprogram://<action>/<path>[?query]`. It deliberately validates the raw
 * components before any WHATWG URL normalization so encoded dot segments or
 * separators cannot change the page route passed to a host mini-program.
 */
export function parseMiniProgramUri(value: string): MiniProgramUri | null {
    const raw = value.trim();
    if (raw.slice(0, PREFIX.length).toLowerCase() !== PREFIX) return null;
    if (raw.includes('#')) return null;

    const hierarchyAndQuery = raw.slice(PREFIX.length);
    const queryIndex = hierarchyAndQuery.indexOf('?');
    const hierarchy = queryIndex >= 0
        ? hierarchyAndQuery.slice(0, queryIndex)
        : hierarchyAndQuery;
    const rawQuery = queryIndex >= 0
        ? hierarchyAndQuery.slice(queryIndex + 1)
        : null;

    const pathIndex = hierarchy.indexOf('/');
    if (pathIndex <= 0) return null;

    const rawAction = hierarchy.slice(0, pathIndex).toLowerCase();
    if (!ACTIONS.has(rawAction as MiniProgramUriAction)) return null;

    const rawPath = hierarchy.slice(pathIndex);
    if (rawPath === '/' || !rawPath.startsWith('/')) return null;
    if (rawQuery !== null && !QUERY.test(rawQuery)) return null;

    const segments = rawPath.slice(1).split('/');
    if (segments.some((segment) => !PATH_SEGMENT.test(segment))) return null;

    for (const segment of segments) {
        let decoded: string;
        try {
            decoded = decodeURIComponent(segment);
        } catch {
            return null;
        }
        if (
            decoded === '.'
            || decoded === '..'
            || decoded.includes('/')
            || decoded.includes('\\')
            || decoded.includes('\0')
        ) {
            return null;
        }
    }

    return {
        action: rawAction as MiniProgramUriAction,
        route: `${rawPath}${rawQuery === null ? '' : `?${rawQuery}`}`,
    };
}

export function isMiniProgramUri(value: string): boolean {
    return value.trim().slice(0, `${MINI_PROGRAM_URI_SCHEME}:`.length).toLowerCase()
        === `${MINI_PROGRAM_URI_SCHEME}:`;
}
